src/Middleware/HttpLogging/src/FileLoggerProcessor.cs

```diff
@@ -0,0 +1,262 @@
+            _outputTask = Task.Run(ProcessLogQueue);
+        }
+
+        public void EnqueueMessage(string message)
+        {
+            if (!_messageQueue.IsAddingCompleted)
+            {
+                try
+                {
+                    _messageQueue.Add(message);
+                    return;
+                }
+                catch (InvalidOperationException) { }
+            }
+        }
+
+        private async Task ProcessLogQueue()
+        {
+            while (!_cancellationTokenSource.IsCancellationRequested)
+            {
+                while (_messageQueue.TryTake(out var message))
+                {
+                    _currentBatch.Add(message);
+                }
+                if (_currentBatch.Count > 0)
+                {
+                    try
+                    {
+                        await WriteMessagesAsync(_currentBatch, _cancellationTokenSource);
+                    }
+                    catch (Exception ex)
+                    {
+                        Log.WriteMessagesFailed(_logger, ex);
+                    }
+
+                    _currentBatch.Clear();
+                }
+                else
+                {
+                    try
+                    {
+                        await Task.Delay(_flushInterval, _cancellationTokenSource.Token);
+                    }
+                    catch
+                    {
+                        // Exit if task was canceled
+                        return;
+                    }
+                }
+            }
+        }
+
+        private async Task WriteMessagesAsync(List<string> messages, CancellationTokenSource cancellationTokenSource)
+        {
+            // Files are grouped by day, and written up to _maxFileSize before rolling to a new file
+            DateTime today = DateTime.Now;
+            var fullName = GetFullName(today);
+            var fileInfo = new FileInfo(fullName);
+            if (!TryCreateDirectory())
+            {
+                // return early if we fail to create the directory
+                return;
+            }
+            var streamWriter = GetStreamWriter(fullName);
+
+            try
+            {
+                foreach (var message in messages)
+                {
+                    if (cancellationTokenSource.IsCancellationRequested)
+                    {
+                        return;
+                    }
+                    fileInfo.Refresh();
+                    // Roll to new file if _maxFileSize is reached
+                    // _maxFileSize could be less than the length of the file header - in that case we still write the first log message before rolling
+                    if (fileInfo.Exists && fileInfo.Length > _maxFileSize)
+                    {
+                        _fileNumber++;
+                        fullName = GetFullName(today);
+                        fileInfo = new FileInfo(fullName);
+                        streamWriter.Dispose();
```

Comment:
Double dispose if the following `TryCreateDirectory` returns false

Reply:
https://github.com/dotnet/aspnetcore/pull/33251/commits/83050fd92fcf06353a109f83f27e0ce1430bc5d5